### PR TITLE
feat: add redis registry

### LINKED ISSUE #23: Proposal: Support redis registry
**Is your feature request related to a problem? Please describe.**

Currently, there is no redis registry support, adding it can enrich user choices.

**Describe the solution you'd like**

Support redis registry.

### PR DESCRIPTION
**Pull request by justlorain**
#### What type of PR is this?

feat: A new feature

#### What this PR does / why we need it (English/Chinese):

en: Support redis registry.
zh: 添加redis作为注册中心

#### Which issue(s) this PR fixes:

Fixes #23

### CODE REVIEW

**Comment by li-jin-gou** [comment]
ci need to resolve

**Comment by L2ncE** [comment]
add license https://github.com/hertz-contrib/registry/tree/main/licenses

File: redis/resolver.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/go-redis/redis/v8"
+)
+
+var _ discovery.Resolver = (*redisResolver)(nil)
+
+type redisResolver struct {
+	client *redis.Client
+}
+
+// NewRedisResolver create a redis resolver
```
Comment (L2ncE) [inline redis/resolver.go:33]: create => creates

Comment (justlorain) [reply redis/resolver.go:33]: done, thx

**Comment by justlorain** [comment]
> add license https://github.com/hertz-contrib/registry/tree/main/licenses

done, thx

File: redis/resolver.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
+}
+
+func (r *redisResolver) Target(_ context.Context, target *discovery.TargetInfo) string {
+	return target.Host
+}
+
+func (r *redisResolver) Resolve(ctx context.Context, desc string) (discovery.Result, error) {
+	rdb := r.client
+	fvs := rdb.HGetAll(ctx, fmt.Sprintf("/%s/%s/%s", hertz, desc, server)).Val()
```
Comment (zstone12) [inline redis/resolver.go:51]: 这块是否可以抽出一个generateKey的函数

File: redis/resolver.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	var (
+		ri  registryInfo
+		its []discovery.Instance
+	)
+	for f, v := range fvs {
+		err := json.Unmarshal([]byte(v), &ri)
+		if err != nil {
+			hlog.Warnf("HERTZ: fail to unmarshal with err: %v, ignore instance Addr: %v", err, f)
```
Comment (zstone12) [inline redis/resolver.go:59]: 这里是否需要`continue` 我理解json解析失败，后面的流程有可能都存在问题.

Comment (zstone12) [inline redis/resolver.go:59]: 这里是否需要`continue` 我理解json解析失败，后面的流程有可能都存在问题.

Comment (justlorain) [reply redis/resolver.go:58]: 确实缺少了，已修改，thx

File: redis/resolver.go (hunk 2 above)
Comment (justlorain) [reply redis/resolver.go]: done, thx

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -41,6 +41,11 @@ jobs:
         image: 'xdockerh/eureka-server:latest'
         ports:
           - "8761:8761"
+      redis:
```
Comment (li-jin-gou) [inline .github/workflows/tests.yml:44]: data race need to solve.

**Review by li-jin-gou — CHANGES_REQUESTED** [review]

File: .github/workflows/tests.yml (hunk 4 above)
Comment (justlorain) [reply .github/workflows/tests.yml:44]: got it, thx

Comment (justlorain) [reply .github/workflows/tests.yml:44]: done

**Comment by zstone12** [comment]
lgtm

**Review by zstone12 — APPROVED** [review]

File: redis/redis_test.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	ctx      = context.Background()
+)
+
+func init() {
+	rdb := redis.NewClient(&redis.Options{Addr: "127.0.0.1:6379"})
+	redisCli = rdb
+}
+
+// TestRegister Test the Registry in registry.go
+func TestRegister(t *testing.T) {
```
Comment (li-jin-gou) [inline redis/redis_test.go:39]: The code is OK, but I find that the single test is not combined with Hertz to the extent that I'm not sure it's guaranteed to work with Hertz?

**Review by li-jin-gou — CHANGES_REQUESTED** [review]

File: redis/redis_test.go (hunk 5 above)
Comment (justlorain) [reply redis/redis_test.go:39]: I will enrich the tests combined with hertz, thanks for the reminder.

Comment (justlorain) [reply redis/redis_test.go:47]: done

**Review by li-jin-gou — APPROVED** [review]

**Pull request merged**